pkg/services/libraryelements/database.go

```diff
@@ -243,11 +233,40 @@ func (l *LibraryElementService) deleteLibraryElement(c context.Context, signedIn
 			}
 		}
 
-		// Delete any hanging/invalid connections
-		if _, err = session.Exec(deleteInvalidConnections, element.ID); err != nil {
+		dashboardIDs := []int64{}
+		// get all connections for this element
+		if err := session.SQL("SELECT connection_id FROM library_element_connection where element_id = ?", element.ID).Find(&dashboardIDs); err != nil {
+			return err
+		}
+
+		// then find the dashboards that were supposed to be connected to this element
+		dashs, err := l.dashboardsService.FindDashboards(c, &dashboards.FindPersistedDashboardsQuery{
+			OrgId:                   signedInUser.GetOrgID(),
+			DashboardIds:            dashboardIDs,
+			SkipAccessControlFilter: true, // a user may be able to delete a library element but not read all dashboards. We still need to run this check, so we don't allow deleting elements if dashboards are connected
+		})
+		if err != nil {
+			return err
+		}
+
+		foundDashboardsClause := ""
+		if len(dashs) > 0 {
+			foundDashboardsClause = "AND connection_id NOT IN ("
+			for i, d := range dashs {
+				if i > 0 {
+					foundDashboardsClause += ","
+				}
+				foundDashboardsClause += fmt.Sprintf("%d", d.ID)
+			}
+			foundDashboardsClause += ")"
```

Comment:
Suggestion: I really don't like to use the plain text in the SQL statement. Could we make it a prepared statement and just pass the IDs to the query? So basically building the where clause with `?` and just appending the ids to a slice. Then doing something like 

```golang
		if _, err = session.Exec(`DELETE FROM library_element_connection WHERE element_id = ? `+foundDashboardsClause, element.ID, ids...); err != nil {
			return err
		}
```

Reply:
Good point - unfortunately session.Exec doesn't allow for expanding arrays, unless it is the only parameter, but switched over to using a builder and it is a lot cleaner.